File: app/src/main/java/me/saket/dank/ui/preferences/gestures/GesturePreferenceUiModel.java
package me.saket.dank.ui.preferences.gestures;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.List;

public interface GesturePreferenceUiModel {

  enum Type {
    SUBMISSION_PREVIEW,
    SECTION_HEADER,
    SWIPE_ACTION,
  }

  long adapterId();

  Type type();

  interface ChildAdapter<T extends GesturePreferenceUiModel, VH extends RecyclerView.ViewHolder> {
    VH onCreateViewHolder(LayoutInflater inflater, ViewGroup parent);

    void onBindViewHolder(VH holder, T uiModel);

    void onBindViewHolder(VH holder, T uiModel, List<Object> payloads);
  }
}

File: app/src/main/java/me/saket/dank/ui/preferences/gestures/GesturePreferenceUiModel.kt
package me.saket.dank.ui.preferences.gestures

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

interface GesturePreferenceUiModel {

  enum class Type {
    SUBMISSION_PREVIEW,
    SECTION_HEADER,
    SWIPE_ACTION,
    SWIPE_ACTION_PLACEHOLDER
  }

  fun adapterId(): Long

  fun type(): Type

  interface ChildAdapter<T : GesturePreferenceUiModel, VH : RecyclerView.ViewHolder> {
    fun onCreateViewHolder(inflater: LayoutInflater, parent: ViewGroup): VH

    fun onBindViewHolder(holder: VH, uiModel: T)

    fun onBindViewHolder(holder: VH, uiModel: T, payloads: List<Any>)
  }
}
